### PR TITLE
[feature][broker] Allow to configure the entry filters per namespace and per topic

### LINKED ISSUE #16870: Entry Filters: allow to configure the filters per namespace and not only globally
In PIP-105 we introduced Server side filters.
Now in 2.10 and 2.11 you can only configure the filters (entryFilterNames) in broker.conf.

A natural extension is to let the filters be configured per namespace.

Main reasons:
- different users/tenant may need different filters
- activating a filter on namespaces that don't need the filter is a waste of resources

Therefore we need a API to report to clients that a filter is present on the server, because this way the client may decide to perform some processing locally or let the broker do the processing.

Feature request:
- add per-namespace filters (or even per-topic)
- add an API to get the filters configured on the namespace (accessible with the same permissions needed to consume from the topic/namespace)

### PR DESCRIPTION
**Pull request by gaozhangmin**
Fixes #16870 

### Motivation
Allow to configure the filters per namespace and per topic

### Modifications
Support config namespace and topic level entry filter policy through admin api.
```
pulsar-admin namespaces set-entry-filters
pulsar-admin namespaces get-entry-filters
pulsar-admin namespaces remove-entry-filters

pulsar-admin topics set-entry-filters
pulsar-admin topics get-entry-filters
pulsar-admin topics remove-entry-filters
```
### Verifying this change

org.apache.pulsar.broker.service.plugin.FilterEntryTest
org.apache.pulsar.broker.service.AbstractBaseDispatcherTest#testFilterEntriesForConsumerOfEntryFilter
org.apache.pulsar.broker.admin.AdminApi2Test#testSetNamespaceEntryFilters
org.apache.pulsar.broker.admin.AdminApi2Test#testSetTopicLevelEntryFilters

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java [hunk 1]
```diff
@@ -2651,5 +2652,66 @@ public void finished(int total, int errors, int unknown) throws Exception {
...
+    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
+            @ApiResponse(code = 404, message = "Namespace does not exist") })
+    public void getEntryFiltersPerTopic(
+            @Suspended final AsyncResponse asyncResponse,
+            @PathParam("tenant") String tenant,
+            @PathParam("namespace") String namespace) {
+        validateNamespaceName(tenant, namespace);
+        validateNamespacePolicyOperationAsync(namespaceName, PolicyName.ENTRY_FILTERS, PolicyOperation.READ)
+                .thenCompose(__ -> getNamespacePoliciesAsync(namespaceName))
+                .thenAccept(polices -> asyncResponse.resume(polices.entryFilters))
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java:2667]: we must not  return the entry filters directory to clients, as it will disclose sensitive information about the broker filesystem

please create a new java class to model the response and report only the list of enabled entry filters

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java [hunk 2]
```diff
@@ -2651,5 +2652,66 @@ public void finished(int total, int errors, int unknown) throws Exception {
...
+
+    @POST
+    @Path("/{tenant}/{namespace}/entryFilters")
+    @ApiOperation(value = "Set entry filters for namespace")
+    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
+            @ApiResponse(code = 404, message = "Tenant or cluster or namespace doesn't exist")})
+    public void setEntryFiltersPerTopic(@Suspended AsyncResponse asyncResponse, @PathParam("tenant") String tenant,
+                                       @PathParam("namespace") String namespace,
+                                       @ApiParam(value = "entry filters", required = true)
+                                               EntryFilters entryFilters) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java:2684]: the client must not be able to set the directory, but only the list of enabled filters

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 3]
```diff
@@ -321,9 +322,6 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
                 .newSingleThreadScheduledExecutor(new DefaultThreadFactory("pulsar-stats-updater"));
         this.authorizationService = new AuthorizationService(
                 pulsar.getConfiguration(), pulsar().getPulsarResources());
-        if (!pulsar.getConfiguration().getEntryFilterNames().isEmpty()) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:324]: it is better still keep a list of global entry filters and initialise them only once.

in a broker with thousands of topics the overhead may be non negligible, a EntryFilterFactory may have to perform some initialisation at boot  

also a EntryFilter may have some local cache.
in 2.10 the EntryFilter instance is global, so the cache is shared, if we create a filter entry per each topic then this kind of caching will be useless.

I think that we have to keep a list of EntryFilters globally, per-namespace and per-topic.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/AbstractBaseDispatcherTest.java [hunk 4]
```diff
@@ -214,6 +211,10 @@ protected boolean isConsumersExceededOnSubscription() {
 
         @Override
         public boolean trackDelayedDelivery(long ledgerId, long entryId, MessageMetadata msgMetadata) {
+            if (!msgMetadata.hasDeliverAtTime()) {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/AbstractBaseDispatcherTest.java:214]: this change seems unrelated to the patch

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/EntryFilters.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * The class name for the entry filter.
+     */
+    private String entryFilterNames;
+
+    /**
+     * The directory for all the entry filter implementations.
+     */
+    private String entryFiltersDirectory;
```
Comment (eolivelli) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/EntryFilters.java:42]: This is a parameter that is only useful only in ServiceConfiguration (broker.conf) because it refers to the local filesystem of every specific broker.

I think we should drop this from here

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/EntryFilters.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class EntryFilters {
+    /**
+     * The description of the entry filter to be used for user help.
+     */
+    private String description;
```
Comment (eolivelli) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/EntryFilters.java:32]: this is not useful, we generally do not store "descriptions". we have to drop this field

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopicPolicies.java [hunk 7]
```diff
@@ -146,6 +147,75 @@ public CmdTopicPolicies(Supplier<PulsarAdmin> admin) {
...
+        private String  description = "";
+
+        @Parameter(names = { "--entry-filters-name", "-efn" },
+                description = "The class name for the entry filter.", required = true)
+        private String  entryFiltersName = "";
+
+
+        @Parameter(names = { "--entry-filters-dir", "-efd" },
+                description = " The directory for all the entry filter implementations.", required = true)
+        private String  entryFiltersDirectory = "";
```
Comment (eolivelli) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopicPolicies.java:191]: as in other places please drop this field

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopicPolicies.java [hunk 8]
```diff
@@ -146,6 +147,75 @@ public CmdTopicPolicies(Supplier<PulsarAdmin> admin) {
...
+    }
+
+    @Parameters(commandDescription = "Set entry filters for a topic")
+    private class setEntryFiltersPerTopic extends CliCommand {
+        @Parameter(description = "persistent://tenant/namespace/topic", required = true)
+        private java.util.List<String> params;
+
+        @Parameter(names = { "--desc", "-d" },
+                description = "The description of the entry filter to be used for user help.", required = false)
+        private String  description = "";
```
Comment (eolivelli) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopicPolicies.java:182]: as in other places please drop this field

**Review by eolivelli — CHANGES_REQUESTED** [review]
This is a great work!
thank you very much for providing this implementation.

I have left some feedback:
- do not set the directory, it is broker specific (and not all the brokers may have the same filesystem layout)
- do not add a description, we never do that in Pulsar, and it is for very low benefit

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 3 above)
Comment (gaozhangmin) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:324]: I remove this, because it's not used any more.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/AbstractBaseDispatcherTest.java (hunk 4 above)
Comment (gaozhangmin) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/AbstractBaseDispatcherTest.java:214]: it relates to the entry filters unit test.

Comment (gaozhangmin) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/AbstractBaseDispatcherTest.java:214]: I will submit a new pr to solve this.

**Review by eolivelli — CHANGES_REQUESTED** [review]
I have two questions.

It seems to me that with this patch the entryFilters cannot be configured in broker.conf anymore, is this correct ? (in that case we have to restore such support, otherwise existing users that upgrade will have problems)

Ideally I would like to see that in a topic the entryFilters that are applied are the sum of: per-broker entry filters + per-namespace entry filters and per-topic entry filters.

This way in a multi-tenant environment the administrator can have control over what is happening.

Also we should add a flag to allow setting per-namespace and per-topic entry filters.

in some environments the system administrator doesn't want to let the users activate custom behaviour to the users (tenant/namespace operators)

**Comment by eolivelli** [comment]
@gaozhangmin we can chat on slack if you want

**Comment by gaozhangmin** [comment]
@Jason918  PTAL, Thx

**Comment by gaozhangmin** [comment]
> entryFilters cannot be configured in broker.conf anymore, is this correct

1、it's still supported configuring filters via broker.conf, default value is initialized in 
 org.apache.pulsar.broker.service.AbstractTopic#updateTopicPolicyByBrokerConfig
 2、Reference other topic level  policies, I don't think we need a flag there.
 @eolivelli

**Comment by eolivelli** [comment]
@gaozhangmin thanks for your clarification. it makes sense to me.

I have two remaining points:
- I don't find tests about configuring the filters "per broker"
- I do think that the filters must be "additive" and the namespace/topic filters should not override the broker filters

The broker filters may have been set by the system administrator to enforce some rules and if a user (tenant admin?) is able to override the list of filters that that would be some kind of security hole.

If you feel strong that we should keep the behaviour of this patch, then I would ask you to add a configuration parameter to make this behaviour configurable, in order to allow system administrators to enforce some entry filters

**Comment by gaozhangmin** [comment]
@eolivelli  I will add a configuration  parameter in case of the overridden

**Comment by eolivelli** [comment]
@gaozhangmin  thank you for adding the parameter.
The patch now looks generally good.

One last comment: we are missing tests for the new option: we need tests that validate that the option is really working.
After adding the tests I am +1 to this patch, great work !

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/EntryFilterSupport.java [hunk 9]
```diff
@@ -39,9 +40,21 @@ public class EntryFilterSupport {
...
         this.subscription = subscription;
-        if (subscription != null && subscription.getTopic() != null && MapUtils.isNotEmpty(subscription.getTopic()
-                .getBrokerService().getEntryFilters())) {
-            this.entryFilters = subscription.getTopic().getBrokerService().getEntryFilters().values().asList();
+        if (subscription != null && subscription.getTopic() != null) {
+            if (MapUtils.isNotEmpty(subscription.getTopic()
+                    .getBrokerService().getEntryFilters())
+                    && !subscription.getTopic().getBrokerService().pulsar()
+                    .getConfiguration().isAllowOverrideEntryFilters()) {
+                this.entryFilters = subscription.getTopic().getBrokerService().getEntryFilters().values().asList();
```
Comment (gaozhangmin) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/EntryFilterSupport.java:48]: @eolivelli  Here is `isAllowOverrideEntryFilters `

**Comment by gaozhangmin** [comment]
> option

@eolivelli  What you mean  for new option is the configuration `allowOverrideEntryFilters`, Right?

**Comment by eolivelli** [comment]
> @eolivelli What you mean for new option is the configuration `allowOverrideEntryFilters`, Right?

correct

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java [hunk 10]
```diff
@@ -71,6 +76,57 @@ protected void cleanup() throws Exception {
         internalCleanup();
     }
 
+    @Test
+    public void testOverride() throws Exception {
```
Comment (gaozhangmin) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:80]: @eolivelli  test added. PTAL

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java [hunk 11]
```diff
@@ -87,20 +87,20 @@
  */
 public abstract class MockedPulsarServiceBaseTest extends TestRetrySupport {
     public final static String BROKER_KEYSTORE_FILE_PATH =
-            Resources.getResource("certificate-authority/jks/broker.keystore.jks").getPath();
+            "";
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java:90]: Is this change related to this patch?

**Review by eolivelli — APPROVED** [review]
LGTM

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java (hunk 11 above)
Comment (gaozhangmin) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java:90]: I have reverted this change.
